consensus/src/consensus_observer/observer/consensus_observer.rs

```diff
@@ -817,6 +843,98 @@ impl ConsensusObserver {
+    ) {
+        // If execution pool is disabled, ignore the message
+        let execution_pool_window_size = match self.get_execution_pool_window_size() {
+            Some(window_size) => window_size,
+            None => {
+                // Log the failure and update the invalid message counter
+                warn!(
+                    LogSchema::new(LogEntry::ConsensusObserver).message(&format!(
+                        "Received ordered block with window message but execution pool is disabled! Ignoring: {:?}",
+                        ordered_block_with_window.ordered_block().proof_block_info()
+                    ))
+                );
+                update_invalid_message_counter(
+                    &peer_network_id,
+                    metrics::ORDERED_BLOCK_WITH_WINDOW_LABEL,
+                );
+                return;
+            },
+        };
+
+        // Verify the ordered blocks before processing
+        let ordered_block = ordered_block_with_window.ordered_block();
+        if let Err(error) = ordered_block.verify_ordered_blocks() {
+            // Log the error and update the invalid message counter
+            error!(
+                LogSchema::new(LogEntry::ConsensusObserver).message(&format!(
+                    "Failed to verify ordered block with window! Ignoring: {:?}, Error: {:?}",
+                    ordered_block.proof_block_info(),
+                    error
+                ))
+            );
+            update_invalid_message_counter(
+                &peer_network_id,
+                metrics::ORDERED_BLOCK_WITH_WINDOW_LABEL,
+            );
+            return;
+        };
+
+        // Verify the execution pool window contents
+        let execution_pool_window = ordered_block_with_window.execution_pool_window();
+        if let Err(error) = execution_pool_window.verify_window_contents(execution_pool_window_size)
+        {
+            // Log the error and update the invalid message counter
+            error!(
+                LogSchema::new(LogEntry::ConsensusObserver).message(&format!(
+                    "Failed to verify execution pool window contents! Ignoring: {:?}, Error: {:?}",
+                    ordered_block.proof_block_info(),
+                    error
+                ))
+            );
+            update_invalid_message_counter(
+                &peer_network_id,
+                metrics::ORDERED_BLOCK_WITH_WINDOW_LABEL,
+            );
+            return;
+        };
+
+        // Get the epoch and round of the first block
+        let first_block = ordered_block.first_block();
+        let first_block_epoch_round = (first_block.epoch(), first_block.round());
+
+        // Determine if the block is behind the last ordered block, or if it is already pending
+        let last_ordered_block = self.get_last_ordered_block();
+        let block_out_of_date =
+            first_block_epoch_round <= (last_ordered_block.epoch(), last_ordered_block.round());
```

Comment:
Just as a sanity check, last ordered block is the most recent right?

Reply:
Correct 😄